src/sql/src/plan/statement/ddl.rs

```diff
@@ -2995,97 +2995,205 @@ fn kafka_sink_builder(
+            seen: _,
+        } = options.try_into()?;
 
-            let item = scx.get_item_by_resolved_name(&connection)?;
-            let csr_connection = match item.connection()? {
-                Connection::Csr(_) => item.id(),
-                _ => {
-                    sql_bail!(
-                        "{} is not a schema registry connection",
-                        scx.catalog
-                            .resolve_full_name(item.name())
-                            .to_string()
-                            .quoted()
-                    )
-                }
-            };
-            let CsrConfigOptionExtracted {
-                avro_key_fullname,
-                avro_value_fullname,
-                null_defaults,
-                key_doc_options,
-                value_doc_options,
-                key_compatibility_level,
-                value_compatibility_level,
-                seen: _,
-            } = options.try_into()?;
+        if key_desc_and_indices.is_none() && avro_key_fullname.is_some() {
+            sql_bail!("Cannot specify AVRO KEY FULLNAME without a corresponding KEY field");
+        }
 
-            if key_desc_and_indices.is_none() && avro_key_fullname.is_some() {
-                sql_bail!("Cannot specify AVRO KEY FULLNAME without a corresponding KEY field");
-            }
+        if key_desc_and_indices.is_some()
+            && (avro_key_fullname.is_some() ^ avro_value_fullname.is_some())
+        {
+            sql_bail!("Must specify both AVRO KEY FULLNAME and AVRO VALUE FULLNAME when specifying generated schema names");
+        }
 
-            if key_desc_and_indices.is_some()
-                && (avro_key_fullname.is_some() ^ avro_value_fullname.is_some())
-            {
-                sql_bail!("Must specify both AVRO KEY FULLNAME and AVRO VALUE FULLNAME when specifying generated schema names");
-            }
+        let options = AvroSchemaOptions {
+            avro_key_fullname,
+            avro_value_fullname,
+            set_null_defaults: null_defaults,
+            is_debezium: matches!(envelope, SinkEnvelope::Debezium),
+            sink_from: Some(sink_from),
+            value_doc_options,
+            key_doc_options,
+        };
+        Ok((
+            csr_connection,
+            options,
+            key_compatibility_level,
+            value_compatibility_level,
+        ))
+    };
 
-            let options = AvroSchemaOptions {
-                avro_key_fullname,
-                avro_value_fullname,
-                set_null_defaults: null_defaults,
-                is_debezium: matches!(envelope, SinkEnvelope::Debezium),
-                sink_from: Some(sink_from),
-                value_doc_options,
-                key_doc_options,
-            };
+    // Helper function to determine if a relation desc is valid for text or binary format.
+    // We only allow a single column of non-list/map type.
+    let valid_text_binary_desc = |desc: &RelationDesc| {
```

Comment:
This function is written in a way that will silently fail if we ever add a scalar type that is not supported. It should check the inverse of what it currently does. Also, since the closure doesn't capture any variables we can extract it in a standalone function.